floris/tools/energy_ratio.py

```diff
@@ -48,6 +48,283 @@ def _get_confidence_bounds(confidence):
+
+
+
+
+    tuples_unique_base, counts_base = np.unique(base_tuples,axis=0, return_counts=True)
+    tuples_unique_con, counts_con = np.unique(con_tuples,axis=0, return_counts=True)
+    total_counts = counts_base + counts_con
+
+    # Make the weights per tuple
+    weights_base = counts_con.astype(float) / total_counts.astype(float)
+    weights_con = counts_base.astype(float) / total_counts.astype(float)
+
+
+    # Make LUT for each
+    # weights_base_LUT = dict(zip(tuples_unique_base,weights_base ))
+    # weights_con_LUT = dict(zip(tuples_unique_con,weights_con ))
+    weights_base_LUT = dict(zip(tuple(map(tuple,tuples_unique_base)),weights_base ))
+    weights_con_LUT = dict(zip(tuple(map(tuple,tuples_unique_con)),weights_con ))
+
+    # Convert arrays back to tuples
+    base_tuples = tuple(map(tuple,base_tuples))
+    con_tuples = tuple(map(tuple,con_tuples))
+
+    # Make weighting arrays
+    weight_array_base = np.array([weights_base_LUT[x] for x in base_tuples])
+    weight_array_con = np.array([weights_con_LUT[x] for x in con_tuples])
+
+    # # Make a weighting array
+    # lut_base = np.zeros(np.max(ws_unique)+1)
+    # lut_base[ws_unique] = weights_base
+    # weight_array_base = lut_base[ws_base]
+    # lut_con = np.zeros(np.max(ws_unique)+1)
+    # lut_con[ws_unique] = weights_con
+    # weight_array_con = lut_con[ws_con]
+
+    # Weighted sums
+    weight_sum_ref_base = np.sum(ref_pow_base * weight_array_base)
+    weight_sum_test_base = np.sum(test_pow_base * weight_array_base)
+    weight_sum_ref_con = np.sum(ref_pow_con * weight_array_con)
+    weight_sum_test_con = np.sum(test_pow_con * weight_array_con)
+
+    # Ratio and diff
+    ratio_base = weight_sum_test_base / weight_sum_ref_base
+    ratio_con = weight_sum_test_con / weight_sum_ref_con
+    ratio_diff = ratio_con - ratio_base
+    p_change = 100. * ratio_diff / ratio_base
+
+    # Get the counts
+    counts_base = len(ref_pow_base)
+    counts_con = len(ref_pow_con)
+    counts_diff = np.min([counts_base, counts_con])
+    counts_pchange = counts_diff
+
+    # Also compute power gain and energy gain in raw units
+
+
+    return ratio_base, ratio_con, ratio_diff, p_change, counts_base, counts_con, counts_diff, counts_pchange
+
+def calculate_balanced_energy_ratio_region(reference_power_baseline,
+                                    test_power_baseline,
+                                    wind_speed_array_baseline,
+                                    wind_direction_array_baseline,
+                                    reference_power_controlled,
+                                    test_power_controlled,
+                                    wind_speed_array_controlled,
+                                    wind_direction_array_controlled,
+                                    confidence=95,
+                                    n_boostrap=None
+                                    ):
+    """
+    Calculate a balanced energy ratio over a region
```

Comment:
doc string should be updated. Example: get rid of wind_direction_bins description